File: Stage1st/Scene/Reply/MahjongFaceInputView/NSAttributedString+MahjongFaceExtension.m
//
//  NSAttributedString+MahjongFaceExtension.h
//  Stage1st
//
//  Created by Zheng Li on 5/19/15.
//  Copyright (c) 2015 Renaissance. All rights reserved.
//

#import <UIKit/UIKit.h>
#import "NSAttributedString+MahjongFaceExtension.h"
#import "S1MahjongFaceTextAttachment.h"

@implementation NSAttributedString (MahjongFaceExtension)

- (NSString *)getPlainString {
    NSMutableString *plainString = [NSMutableString stringWithString:self.string];
    __block NSUInteger base = 0;
    
    [self enumerateAttribute:NSAttachmentAttributeName inRange:NSMakeRange(0, self.length) options:0 usingBlock:^(id value, NSRange range, BOOL *stop) {
        if (value && [value isKindOfClass:[S1MahjongFaceTextAttachment class]]) {
            [plainString replaceCharactersInRange:NSMakeRange(range.location + base, range.length)
                                       withString:((S1MahjongFaceTextAttachment *) value).mahjongFaceTag];
            base += ((S1MahjongFaceTextAttachment *) value).mahjongFaceTag.length - 1;
        }
    }];
    
    return plainString;
}

@end

File: Stage1st/Scene/Reply/MahjongFaceInputView/NSAttributedString+MahjongFaceExtension.swift
//
//  NSAttributedString+MahjongFaceExtension.swift
//  Stage1st
//
//  Created by Zheng Li on 2019/1/6.
//  Copyright © 2019 Renaissance. All rights reserved.
//

import UIKit

extension NSAttributedString {

    @objc func s1_getPlainString() -> String {
        var plainString = self.string as NSString
        var base = 0

        self.enumerateAttribute(.attachment, in: NSRange(location: 0, length: length), options: []) { (value, range, stop) in
            guard let value = value as? MahjongFaceTextAttachment else {
                return
            }

            plainString = plainString.replacingCharacters(in: NSRange(location: range.location + base, length: range.length), with: value.tag) as NSString
            base += value.tag.count - 1
        }
        return plainString as String
    }
}
